File: client-samples/ios-visionos/StreamKit/Sources/StreamKit/CameraPreviewView.swift
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

/*
 * StreamKit — CameraPreviewView
 *
 * SwiftUI view that renders the local camera track of a StreamSession.
 *
 * Encapsulates the LiveKit `SwiftUIVideoView` so application code does not
 * need to import the LiveKit SDK.  Renders a transparent view when the
 * camera is not active or the session is backed by a non-LiveKit transport.
 */

import Combine
import LiveKit
import SwiftUI

// MARK: - CameraPreviewView

/// Renders the local camera feed published by a ``StreamSession``.
///
/// Place this view anywhere in your layout to give the user a "what the
/// camera sees" preview that matches the web client's `<video>` element.
/// The view is empty (transparent) when ``StreamSession/localCameraTrack``
/// is `nil` (camera stopped, or non-LiveKit backend).
///
/// Pair with ``CameraPreviewAspect`` to size the surrounding container to
/// the actual capture aspect ratio (so portrait phone capture renders as
/// 9:16, landscape sensors as 16:9):
///
/// ```swift
/// @StateObject private var aspect = CameraPreviewAspect()
///
/// CameraPreviewView(session: model.session)
///     .aspectRatio(aspect.value(default: 16.0 / 9.0), contentMode: .fit)
///     .onAppear { aspect.attach(to: model.session) }
/// ```
public struct CameraPreviewView: View {
    private let session: StreamSession?

    public init(session: StreamSession?) {
        self.session = session
    }

    public var body: some View {
        if let track = session?.localCameraTrack {
            // LiveKit's SwiftUIVideoView handles mirroring for front-facing
            // capture and resizes the underlying RTC video sink as the
            // SwiftUI layout changes.
            SwiftUIVideoView(track, layoutMode: .fill, mirrorMode: .auto)
        } else {
            Color.clear
        }
    }
}

// MARK: - CameraPreviewAspect

/// Observes the live camera track's frame dimensions and exposes them as
/// an aspect ratio (`width / height`) for SwiftUI layout.
///
/// Returns `nil` until the first frame has been published by LiveKit; pair
/// it with a sensible fallback so the layout is stable before capture starts:
///
/// ```swift
/// @StateObject private var aspect = CameraPreviewAspect()
///
/// var body: some View {
///     CameraPreviewView(session: model.session)
///         .aspectRatio(aspect.value(default: 16.0 / 9.0), contentMode: .fit)
///         .onChange(of: model.session?.localCameraTrack === nil) { _, _ in
///             aspect.attach(to: model.session)
///         }
/// }
/// ```
///
/// Avoids leaking LiveKit types into app code — the only public surface is
/// a `CGFloat` aspect ratio.
@MainActor
public final class CameraPreviewAspect: ObservableObject {
    @Published public private(set) var aspectRatio: CGFloat?

    private var observer: TrackDelegateObserver?
    private var cancellable: AnyCancellable?

    public init() {}

    /// Returns the live aspect ratio if known, otherwise `defaultValue`.
    public func value(default defaultValue: CGFloat) -> CGFloat {
        aspectRatio ?? defaultValue
    }

    /// Attaches to the session's current camera track. Call this whenever
    /// the camera is started, stopped, or switched (front ↔ back).
    public func attach(to session: StreamSession?) {
        cancellable = nil
        observer = nil
        aspectRatio = nil

        guard let track = session?.localCameraTrack else { return }

        let observer = TrackDelegateObserver(track: track)
        self.observer = observer

        update(from: observer.dimensions)
        cancellable = observer.$dimensions.sink { [weak self] dimensions in
            Task { @MainActor [weak self] in
                self?.update(from: dimensions)
            }
        }
    }

    private func update(from dimensions: Dimensions?) {
        guard let dimensions, dimensions.width > 0, dimensions.height > 0 else {
            aspectRatio = nil
            return
        }
        aspectRatio = CGFloat(dimensions.width) / CGFloat(dimensions.height)
    }
}

File: client-samples/android/app/src/main/java/com/nvidia/xrai/streamkitsample/streamkit/ui/CameraPreviewView.kt
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

/*
 * StreamKit — CameraPreviewView
 *
 * Jetpack Compose view that renders the local camera track of a
 * StreamSession.  Wraps LiveKit's TextureViewRenderer in an AndroidView so
 * application code does not have to depend on the LiveKit SDK directly.
 */

package com.nvidia.xrai.streamkitsample.streamkit.ui

import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.viewinterop.AndroidView
import com.nvidia.xrai.streamkitsample.streamkit.StreamSession
import io.livekit.android.renderer.TextureViewRenderer
import livekit.org.webrtc.VideoSink

/**
 * Renders the local camera feed published by [session].
 *
 * Place anywhere in your layout to give the user a "what the camera sees"
 * preview matching the web client's `<video>` element.  The view is empty
 * when [StreamSession.localCameraTrack] is `null` (camera stopped, or
 * non-LiveKit backend).
 *
 * Pair with [rememberCameraPreviewAspectRatio] to size the surrounding
 * container to the actual capture aspect ratio (so portrait phone capture
 * renders as 9:16, landscape sensors as 16:9):
 *
 * ```kotlin
 * val aspect = rememberCameraPreviewAspectRatio(session) ?: (16f / 9f)
 * Box(Modifier.aspectRatio(aspect)) {
 *     CameraPreviewView(session = session)
 * }
 * ```
 */
@Composable
fun CameraPreviewView(
    session: StreamSession?,
    modifier: Modifier = Modifier,
) {
    val context = LocalContext.current
    val track = session?.localCameraTrack

    // Renderer lifetime is keyed to the track instance — recreate (and
    // release) the TextureViewRenderer whenever the track identity changes
    // so the addRenderer/removeRenderer pair always operates on a live
    // renderer. Track-less compositions still execute the hooks below to
    // keep the slot table stable across null ↔ non-null transitions.
    val renderer = remember(context, session, track) {
        if (track == null) null
        else TextureViewRenderer(context).also { session.initVideoRenderer(it) }
    }

    DisposableEffect(track, renderer) {
        if (track != null && renderer != null) {
            track.addRenderer(renderer)
        }
        onDispose {
            if (track != null && renderer != null) {
                track.removeRenderer(renderer)
                renderer.release()
            }
        }
    }

    if (renderer != null) {
        AndroidView(
            factory = { renderer },
            modifier = modifier.fillMaxSize(),
        )
    }
}

/**
 * Observes the live camera track's frame dimensions and returns the latest
 * aspect ratio (`width / height`, rotation-corrected) for Compose layout.
 *
 * Returns `null` until the first frame arrives, so callers should pair it
 * with a sensible fallback (typically 16:9):
 *
 * ```kotlin
 * val aspect = rememberCameraPreviewAspectRatio(session) ?: (16f / 9f)
 * Box(Modifier.aspectRatio(aspect)) {
 *     CameraPreviewView(session = session)
 * }
 * ```
 *
 * Internally attaches a lightweight [VideoSink] to the track that only
 * inspects the frame's rotated dimensions — no decoding or rendering work.
 */
@Composable
fun rememberCameraPreviewAspectRatio(session: StreamSession?): Float? {
    val track = session?.localCameraTrack
    val aspect = remember(track) { mutableStateOf<Float?>(null) }

    DisposableEffect(track) {
        if (track == null) {
            aspect.value = null
            return@DisposableEffect onDispose { }
        }

        val sink = VideoSink { frame ->
            val w = frame.rotatedWidth
            val h = frame.rotatedHeight
            if (w > 0 && h > 0) {
                val newAspect = w.toFloat() / h.toFloat()
                if (aspect.value != newAspect) {
                    aspect.value = newAspect
                }
            }
        }
        track.addRenderer(sink)

        onDispose { track.removeRenderer(sink) }
    }

    return aspect.value
}
